### DISCUSSION TITLE
VRR breakers

### DISCUSSION THREAD

**Discussion by UjinT34** [hyprwm/Hyprland#7386] [General]
### Multi-monitor
Multi-monitor doesn't work with nvidia (vrr activates but refresh rate is stuck at max) https://forums.developer.nvidia.com/t/feature-g-sync-freesync-under-wayland-session/220822/116?page=2. Others should work fine.

### Animations
Any running animation will break VRR.

### Cursor
DIfferent cursor setting combinations might break VRR. Frames scheduled by HL and AQ should be handled by `cursor:no_break_fs_vrr`. The breakage comes from excessive `_CWlSurfaceCommit` calls. Those extra calls are observed when pointer frame is sent with lock constraint active or a HW cursor buffer is present with atomic modesetting.

Games that use cursor lock usually hide the compositor cursor. They might still render cursor themselves. This is still counts as an invisible cursor.
Some games might switch between different modes, e.g. locked invisible during gameplay and unlocked visible in menus.
I don't know any games that might have locked != invisible.
Assuming `cursor:no_break_fs_vrr = true` and `cursor:min_refresh_rate` less than the target framerate:
| HW  | Locked | Invisible | Status |
| :---: | :---: | :---: | ------------- |
| &mdash;  | &mdash;  |  &mdash;  | works |
| &mdash; | &mdash;  |  &check;  | untested |
| &mdash; | &check;  |  &mdash;  | untested |
| &mdash; | &check;  |  &check;  |works with https://github.com/hyprwm/Hyprland/pull/6877 |
| &check;  | &mdash;  |  &mdash;  | breaks VRR |
| &check;  | &mdash;  |  &check;  |untested |
| &check;  | &check;  |  &mdash;  |untested |
| &check;  | &check;  |  &check;  | works with https://github.com/hyprwm/Hyprland/pull/6877 |

Does anyone has a different experience? Are there any wayland compositors that have a working VRR with HW cursors and atomic modesetting?

**Comment by Atemu** [comment]
Thanks a bunch for the info (and the PR!).

How would one deduce whether the cursor is locked or whether they render it themselves?

**Comment by UjinT34** [comment]
With `HYPRLAND_TRACE=1` https://github.com/hyprwm/Hyprland/pull/6877 should output `listener_commitWindow skip` to the log when it skips unwanted frames for a locked cursor. First-person shooters and 3rd-persom games with direct mouse camera control usually use locked hidden cursors. Most of them render the cursor in menus themselves. Some might switch to a visible unlocked cursor. Witcher 3 probably does that.
